sled-agent/src/services.rs

```diff
@@ -1931,43 +1948,33 @@ impl ServiceManager {
 
         // We initialize all the zones we can, but only return one error, if
         // any.
-        let local_existing_zones = Arc::new(Mutex::new(existing_zones));
-        let last_err = Arc::new(Mutex::new(None));
-        stream::iter(requests)
-            // WARNING: Do not use "try_for_each_concurrent" here -- if you do,
-            // it's possible that the future will cancel other ongoing requests
-            // to "initialize_zone".
-            .for_each_concurrent(None, |request| {
-                let local_existing_zones = local_existing_zones.clone();
-                let last_err = last_err.clone();
-                async move {
-                    match self
-                        .initialize_zone(
-                            request,
-                            // filesystems=
-                            &[],
-                        )
-                        .await
-                    {
-                        Ok(running_zone) => {
-                            local_existing_zones.lock().await.insert(
-                                running_zone.name().to_string(),
-                                running_zone,
-                            );
-                        }
-                        Err(err) => {
-                            *last_err.lock().await = Some(err);
-                        }
-                    }
+        let futures = requests.iter().map(|request| {
+            async move {
+                self.initialize_zone(
+                    request,
+                    // filesystems=
+                    &[],
+                )
+                .await
+                .map_err(|error| (request.zone.zone_name(), error))
+            }
+        });
+        let results = futures::future::join_all(futures).await;
+        let mut errors = Vec::new();
+        // TODO: maybe merge and return errors?
```

Comment:
Isn't this basically what you're doing now?

Reply:
Yes haha, this is what happens when you're working till 3am